packages/adapter-utils/src/filter.ts

```diff
@@ -0,0 +1,86 @@
+/*
+*                      Copyright 2021 Salto Labs Ltd.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with
+* the License.  You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+import { Element, Change, PostFetchOptions } from '@salto-io/adapter-api'
+import { types, promises, objects, values } from '@salto-io/lowerdash'
+
+const { concatObjects } = objects
+const { isDefined } = values
+
+// Filters run in a specific order and get a mutable list as input which they may modify
+// to affect the overall result as well as the input for subsequent filters.
+// on preDeploy the filters are run in reverse order and are expected to "undo" any relevant change
+// they made in onFetch. because of this, each filter can expect to get in preDeploy a similar
+// value to what it created in onFetch.
+// onDeploy is called in the same order as onFetch and is expected to do basically the same thing
+// that onFetch does but with a different context (on changes instead of on elements)
+// Each filter will be created once and so it may store context between preDeploy and onDeploy.
+// Note that it cannot store context between onFetch and the other callbacks since these run in
+// separate commands
+export type FilterResult = Record<string, unknown[] | undefined>
+
+export type Filter<T extends FilterResult | void> = Partial<{
+  onFetch(elements: Element[]): Promise<T | void>
+  preDeploy(changes: Change[]): Promise<void>
+  onDeploy(changes: Change[]): Promise<void>
+  onPostFetch(args: PostFetchOptions): Promise<void>
+}>
+
+export type FilterWith<
+  T extends FilterResult | void,
+  M extends keyof Filter<T>
+> = types.HasMember<Filter<T>, M>
+
+export type FilterCreator<
+  R extends FilterResult | void,
+  T,
+> = (opts: T) => Filter<R>
+
+export const filtersRunner = <
+  R extends FilterResult | void,
+  T,
+>(
+    opts: T,
+    filterCreators: ReadonlyArray<FilterCreator<R, T>>,
+  ): Required<Filter<R>> => {
+  // Create all filters in advance to allow them to hold context between calls
+  const allFilters = filterCreators.map(f => f(opts))
+
+  const filtersWith = <M extends keyof Filter<R>>(m: M):
+  FilterWith<R, M>[] =>
+      types.filterHasMember<Filter<R>, M>(m, allFilters)
```

Comment:
indentation nit: the response type having the same indentation is a little confusing (+ adding any new code below this messes up the following line's indentation). maybe
```suggestion
  const filtersWith = <M extends keyof Filter<R>>(m: M): FilterWith<R, M>[] => (
    types.filterHasMember<Filter<R>, M>(m, allFilters)
  )
```